stm/stm.ml

```diff
@@ -569,9 +579,38 @@ end = struct (* {{{ *)
     | Some x -> x
     | None -> raise Vcs_aux.Expired
   let set_state id s =
-    (get_info id).state <- s;
-    if async_proofs_is_master !cur_opt then Hooks.(call state_ready ~doc:dummy_doc (* XXX should be taken in input *) id)
+    let info = get_info id in
+    info.state <- s;
+    let is_full_state_valid = match s with
+      | FullState _ -> true
+      | EmptyState | ErrorState _ | ParsingState _ -> false
+    in
+    if async_proofs_is_master !cur_opt && is_full_state_valid then
+      Hooks.(call state_ready ~doc:dummy_doc (* XXX should be taken in input *) id)
+
   let get_state id = (get_info id).state
+
+  let get_parsing_state id =
+    stm_pperr_endline (fun () -> str "retrieve parsing state state " ++ str (Stateid.to_string id) ++ str " }}}");
+    match (get_info id).state with
+    | FullState s -> s.Vernacstate.parsing
+    | ParsingState s -> s
+    | ErrorState _ | EmptyState -> anomaly Pp.(str"parsing state undefined")
```

Comment:
It would be desirable that this doesn't raise an anomaly, but returns `None`.

Reply:
I think that `EmptyState` should be anomaly (it is an invariant).

While for ErrorState I'm not convinced, in the end, that this is correct. Indeed we want to be able to parse a sentence after a state that was found to be erroneous. In particular this assertion, and the data carried by ErrorState makes "parse;parse;exec;exec" =/= "parse;exec;parse;exec" when the first execution fails, since the following parse is raising an anomaly.

Reply:
Fixed, `ErrorState` now carries a parsing state.